### PR TITLE
[SYCL] Fix compilation with --std=c++20

### PR DESCRIPTION
**Pull request by vladimirlaz**
- add include containing definition of std::bit_cast
 - replace usage of share_pointer::unique() deprecated in C++20
 - add test to verify compilation with different C++ standards

### CODE REVIEW

File: sycl/test/basic_tests/stdcpp_compat.cpp [hunk 1]
```diff
@@ -0,0 +1,7 @@
+// RUN: %clangxx -fsycl -fsycl-unnamed-lambda -fsyntax-only -Xclang -verify -Xclang -verify-ignore-unexpected=note,warning %s -c -o %t.out
```
Comment (bader) [inline sycl/test/basic_tests/stdcpp_compat.cpp:1]:
```suggestion
// RUN: %clangxx -fsycl -fsyntax-only -Xclang -verify -Xclang -verify-ignore-unexpected=note,warning %s -c -o %t.out
```

File: sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp [hunk 2]
```diff
@@ -150,7 +150,7 @@ class __SYCL_EXPORT SYCLMemObjT : public SYCLMemObjI {
 
   __SYCL_DLL_LOCAL void set_final_data_from_storage() {
     MUploadDataFunctor = [this]() {
-      if (!MSharedPtrStorage.unique()) {
+      if (MSharedPtrStorage.use_count() == 1) {
```
Comment (vladimirlaz) [inline sycl/include/CL/sycl/detail/sycl_mem_obj_t.hpp:153]:
```suggestion
      if (MSharedPtrStorage.use_count() > 1) {
```

File: sycl/test/basic_tests/stdcpp_compat.cpp [hunk 3]
```diff
@@ -0,0 +1,7 @@
+// RUN: %clangxx -fsycl -fsyntax-only -Xclang -verify -Xclang -verify-ignore-unexpected=note,warning %s -c -o %t.out
+// RUN: %clangxx -std=c++14 -fsyntax-only -Xclang -verify -I %sycl_include -Xclang -verify-ignore-unexpected=note,warning %s -o -c %t.out
+// RUN: %clangxx -std=c++17 -fsyntax-only -Xclang -verify -I %sycl_include -Xclang -verify-ignore-unexpected=note,warning %s -o -c %t.out
+// RUN: %clangxx -std=c++20 -fsyntax-only -Xclang -verify -I %sycl_include -Xclang -verify-ignore-unexpected=note,warning %s -o -c %t.out
+// expected-no-diagnostics
+
+#include <CL/sycl.hpp>
```
Comment (s-kanaev) [inline sycl/test/basic_tests/stdcpp_compat.cpp:7]: There is also similar file: https://github.com/intel/llvm/blob/sycl/sycl/test/warnings/warnings.cpp.
Is it possible and reasonable to merge these files?

**Review by s-kanaev — APPROVED** [review]

File: sycl/test/basic_tests/stdcpp_compat.cpp (hunk 3 above)
Comment (vladimirlaz) [reply sycl/test/basic_tests/stdcpp_compat.cpp:7]: The tests use similar toolchain  but check different  functionality. We can leave them as different tests.

**Pull request merged**